<a href="https://colab.research.google.com/github/Sayan3sarkar/K-means-Clustering/blob/master/Survival_Rate_Determination_using_K_means_Clustering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# K-means Clustering

We perform <a href="https://www.datascience.com/blog/k-means-clustering">K-means Clustering</a> on  <a href = "https://www.kaggle.com/c/titanic/data">Titanic</a> dataset to determine Survival Rate of passengers. To make this an <a href="https://en.wikipedia.org/wiki/Unsupervised_learning">Unsupervised Learning</a> algorithm, we drop the <b>Survived</b> feature from the Dataset

## Import Required Libraries

In [0]:
import numpy as np
import pandas as pd
from sklearn.cluster import KMeans
from sklearn.preprocessing import MinMaxScaler, LabelEncoder

## Load Dataset

In [2]:
from google.colab import drive
drive.mount('/content/gdrive', force_remount = True)
root_dir = '/content/gdrive/My Drive/'
base_dir = root_dir +'Colab Notebooks/Datasets/Titanic'
base_dir

Mounted at /content/gdrive


'/content/gdrive/My Drive/Colab Notebooks/Datasets/Titanic'

In [3]:
%cd /content/gdrive/My\ Drive/Colab\ Notebooks/Datasets/Titanic

/content/gdrive/My Drive/Colab Notebooks/Datasets/Titanic


In [4]:
#Verifying present working directory
!pwd

/content/gdrive/My Drive/Colab Notebooks/Datasets/Titanic


In [0]:
train = pd.read_csv("train.csv")
test = pd.read_csv("test.csv")

In [6]:
train.head()

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]

In [7]:
test.head()

PassengerId  Pclass  ... Cabin Embarked
0          892       3  ...   NaN        Q
1          893       3  ...   NaN        S
2          894       2  ...   NaN        Q
3          895       3  ...   NaN        S
4          896       3  ...   NaN        S

[5 rows x 11 columns]

## Checking for Missing Values¶

In [8]:
train.isna().head()

PassengerId  Survived  Pclass   Name  ...  Ticket   Fare  Cabin  Embarked
0        False     False   False  False  ...   False  False   True     False
1        False     False   False  False  ...   False  False  False     False
2        False     False   False  False  ...   False  False   True     False
3        False     False   False  False  ...   False  False  False     False
4        False     False   False  False  ...   False  False   True     False

[5 rows x 12 columns]

In [9]:
train.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [11]:
test.isna().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

 ## Performing mean imputation to fill the missing values

In [0]:
train.fillna(train.mean(), inplace = True)

In [0]:
test.fillna(test.mean(), inplace = True)

In [14]:
train.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [15]:
test.isna().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          327
Embarked         0
dtype: int64

## We still find some null values in the dataset. This is because, these values are non-numeric and imputation can be done only on numeric data

## Checking dataset information to figure out unnecessary features

In [16]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            891 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


In [17]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
PassengerId    418 non-null int64
Pclass         418 non-null int64
Name           418 non-null object
Sex            418 non-null object
Age            418 non-null float64
SibSp          418 non-null int64
Parch          418 non-null int64
Ticket         418 non-null object
Fare           418 non-null float64
Cabin          91 non-null object
Embarked       418 non-null object
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


## We don't need $Name, Ticket, Cabin, Embarked$ to decide Survival status. So we drop these fields

In [0]:
train = train.drop(['Name', 'Ticket', 'Cabin', 'Embarked'], 1)

In [0]:
test = test.drop(['Name', 'Ticket', 'Cabin', 'Embarked'], 1)

In [20]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 8 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Sex            891 non-null object
Age            891 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Fare           891 non-null float64
dtypes: float64(2), int64(5), object(1)
memory usage: 55.8+ KB


In [21]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 7 columns):
PassengerId    418 non-null int64
Pclass         418 non-null int64
Sex            418 non-null object
Age            418 non-null float64
SibSp          418 non-null int64
Parch          418 non-null int64
Fare           418 non-null float64
dtypes: float64(2), int64(4), object(1)
memory usage: 22.9+ KB


## We find that $ Sex $ is non-numeric. So we apply <a href = "https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelEncoder.html"> LabelEncoding</a> to the dataset

In [0]:
labelencoder = LabelEncoder()
labelencoder.fit(train['Sex'])
labelencoder.fit(test['Sex'])
train['Sex'] = labelencoder.transform(train['Sex'])
test['Sex'] = labelencoder.transform(test['Sex'])

In [23]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 8 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Sex            891 non-null int64
Age            891 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Fare           891 non-null float64
dtypes: float64(2), int64(6)
memory usage: 55.8 KB


In [24]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 7 columns):
PassengerId    418 non-null int64
Pclass         418 non-null int64
Sex            418 non-null int64
Age            418 non-null float64
SibSp          418 non-null int64
Parch          418 non-null int64
Fare           418 non-null float64
dtypes: float64(2), int64(5)
memory usage: 22.9 KB


## We drop the Survived coloumn and prepare our model

In [0]:
X = np.array(train.drop(['Survived'], 1).astype(float))
y = np.array(train['Survived'])

In [29]:
kmeans = KMeans(n_clusters = 2)
kmeans.fit(X)

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
       n_clusters=2, n_init=10, n_jobs=None, precompute_distances='auto',
       random_state=None, tol=0.0001, verbose=0)

## Finding accuracy of model on training set

In [30]:
correct = 0
for i in range(len(X)):
  pred = np.array(X[i].astype(float))
  pred = pred.reshape(-1, len(pred))
  prediction = kmeans.predict(pred)
  if prediction[0] == y[i]:
    correct += 1
    
print(correct/len(X))

0.49158249158249157


## We tweak certain parameters but before that, we scale the dataset

In [0]:
sc = MinMaxScaler()
X_scaled = sc.fit_transform(X)

In [38]:
kmeans = KMeans(n_clusters = 2, max_iter = 600, algorithm = 'auto', random_state = 66)
kmeans.fit(X_scaled)

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=600,
       n_clusters=2, n_init=10, n_jobs=None, precompute_distances='auto',
       random_state=66, tol=0.0001, verbose=0)

## Finding accuracy of model on training set

In [39]:
correct = 0
for i in range(len(X)):
    pred = np.array(X[i].astype(float))
    pred = pred.reshape(-1, len(pred))
    prediction = kmeans.predict(pred)
    if prediction[0] == y[i]:
        correct += 1
print(correct/len(X))

0.6262626262626263


## Running predicitions on test set

In [0]:
X_test = np.array(test.astype(float))

In [41]:
kmeans.predict(X_test)

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,

## Finding accuracy of model on training set

In [42]:
correct = 0
for i in range(len(X_test)):
    pred = np.array(X_test[i].astype(float))
    pred = pred.reshape(-1, len(pred))
    prediction = kmeans.predict(pred)
    if prediction[0] == y[i]:
        correct += 1
print(correct/len(X_test))

0.6100478468899522


## We obtain a certain improvement after tuning parameters. Training set accuracy = 62.62% and test set accuracy = 61%